### PR TITLE
gh-110525: Add CAPI tests for `set` and `frozenset` objects

### PR DESCRIPTION
**Pull request by sobolevn**
Heavily inspired by https://github.com/python/cpython/blob/main/Modules/_testcapi/dict.c

* Issue: gh-110525

### CODE REVIEW

File: Lib/test/test_capi/test_set.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        self.assertEqual(new().__class__, frozenset)
+        self.assertEqual(new(), frozenset())
+        self.assertEqual(new((1, 1, 2)), frozenset({1, 2}))
+        with self.assertRaisesRegex(TypeError, 'object is not iterable'):
+            new(object())
+        with self.assertRaisesRegex(TypeError, "unhashable type: 'dict'"):
+            new((1, {}))
+
+    def test_set_size(self):
+        l = _testcapi.set_size
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_set.py]: `l` is not good name. It looks like `1` or `I`. Why not use `size`?

**Review by serhiy-storchaka — COMMENTED** [review]
Nice.

Please test also with non-hashable items, non-set as the first argument, NULLs for any argument. If some call crashes, add a comment.

File: Lib/test/test_capi/test_set.py (hunk 1 above)
Comment (vstinner) [reply Lib/test/test_capi/test_set.py]: I would like to suggest a surprising name: set_size.

File: Lib/test/test_capi/test_set.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+class set_child(set):
+    pass
+
+class frozenset_child(frozenset):
+    pass
+
+
+class TestSetCAPI(unittest.TestCase):
+    def assertImmutable(self, action, *args):
+        self.assertRaises(SystemError, action, frozenset(), *args)
```
Comment (vstinner) [inline Lib/test/test_capi/test_set.py:17]: SystemError? That's a surprising error. Usually, it's used when the C API is misused, like passing NULL or the wrong type. frozenset is a "wrong type" here?

File: Lib/test/test_capi/test_set.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
+import unittest
+
+from test.support import import_helper
+
+# Skip this test if the _testcapi module isn't available.
+_testcapi = import_helper.import_module('_testcapi')
+
+class set_child(set):
```
Comment (vstinner) [inline Lib/test/test_capi/test_set.py]: Usually I call this a "subclass", not a "child". I'm more used to parent/child names when it comes to processes.

File: Lib/test/test_capi/test_set.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        self.assertTrue(check(set()))
+        self.assertTrue(check({1, 2}))
+        self.assertTrue(check(frozenset()))
+        self.assertTrue(check(frozenset({1, 2})))
+        self.assertFalse(check(set_child()))
+        self.assertFalse(check(frozenset_child()))
+        self.assertFalse(check(object()))
+
+    def test_set_new(self):
+        new = _testcapi.set_new
```
Comment (vstinner) [inline Lib/test/test_capi/test_set.py]: I would prefer "set_new" name here, and "frozenset_new" below.

File: Lib/test/test_capi/test_set.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        self.assertEqual(l(frozenset()), 0)
+        self.assertEqual(l({1, 1, 2}), 2)
+        self.assertEqual(l(frozenset({1, 1, 2})), 2)
+        self.assertEqual(l(set_child((1, 2, 3))), 3)
+        self.assertEqual(l(frozenset_child((1, 2, 3))), 3)
+        with self.assertRaises(SystemError):
+            l([])
+
+    def test_set_get_size(self):
+        l = _testcapi.set_get_size
```
Comment (vstinner) [inline Lib/test/test_capi/test_set.py]: I would prefer "set_get_size" or get_size" name.

**Review by vstinner — COMMENTED** [review]
I was never comfortable with _testcapi having fun with changing C API names. We cannot simply use PySet_New() name in _testcapi and in the test? I would make it more obvious that the test is on... PySet_New().

File: Lib/test/test_capi/test_set.py (hunk 2 above)
Comment (sobolevn) [reply Lib/test/test_capi/test_set.py:17]: Yes, immutable frozenset is a wrong type for mutation-based functions :)

**Comment by sobolevn** [comment]
> I was never comfortable with _testcapi having fun with changing C API names. We cannot simply use PySet_New() name in _testcapi and in the test?

I think that this is intentional, because we don't really call `PySet_Check`, we call `set_check` intermediate function which changes some behavior compared to `PySet_Check`. For example, it converts `None` to `NULL` (so we can test passing `NULL` easier).

So, I would prefer to keep this naming convention.

File: Lib/test/test_capi/test_set.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+    def test_set_new(self):
+        set_new = _testcapi.set_new
+        self.assertEqual(set_new().__class__, set)
+        self.assertEqual(set_new(), set())
+        self.assertEqual(set_new((1, 1, 2)), {1, 2})
+        with self.assertRaisesRegex(TypeError, 'object is not iterable'):
+            set_new(object())
+        with self.assertRaisesRegex(TypeError, 'object is not iterable'):
+            set_new(None)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_set.py]: None can conflict with NULL. It is better to use other non-iterable: an integer or `object()`.

File: Lib/test/test_capi/test_set.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    def test_set_get_size(self):
+        get_size = _testcapi.set_get_size
+        self.assertEqual(get_size(set()), 0)
+        self.assertEqual(get_size(frozenset()), 0)
+        self.assertEqual(get_size({1, 1, 2}), 2)
+        self.assertEqual(get_size(frozenset({1, 1, 2})), 2)
+        self.assertEqual(get_size(set_subclass((1, 2, 3))), 3)
+        self.assertEqual(get_size(frozenset_subclass((1, 2, 3))), 3)
+        # CRASHES: get_size(NULL)
+        # CRASHES: get_size(object())
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_set.py:131]: `[]` is used as an example of non-set argument in the previous test. I think that it is better to use the same example in both tests. Or both if the behavior is different.

File: Lib/test/test_capi/test_set.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        for cls in (set, frozenset, set_subclass, frozenset_subclass):
+            with self.subTest(cls=cls):
+                instance = cls((1, 2))
+                self.assertTrue(contains(instance, 1))
+                self.assertFalse(contains(instance, 'missing'))
+                with self.assertRaisesRegex(TypeError, "unhashable type: 'list'"):
+                    contains(instance, [])
+                # CRASHES: contains(instance, NULL)
+                # CRASHES: contains(NULL, object())
+                # CRASHES: contains(NULL, NULL)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_set.py]: These two do not belong to the loop.

File: Lib/test/test_capi/test_set.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+                instance = cls((1, 2))
+                self.assertEqual(add(instance, 1), 0)
+                self.assertEqual(instance, {1, 2})
+                self.assertEqual(add(instance, 3), 0)
+                self.assertEqual(instance, {1, 2, 3})
+                with self.assertRaisesRegex(TypeError, "unhashable type: 'list'"):
+                    add(instance, [])
+                # CRASHES: add(NULL, object())
+                # CRASHES: add(instance, NULL)
+                # CRASHES: add(NULL, NULL)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_set.py]: Move comments that do not depend on `instance` outside of the loop.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. Thank you @sobolevn for your PR.

**Comment by sobolevn** [comment]
Should I do a backport?

**Comment by serhiy-storchaka** [comment]
It would be nice. Otherwise I will do it.

**Pull request merged**